### PR TITLE
Arrows control in Pause/Fail overlays

### PR DESCRIPTION
**Pull request by EVAST9919**

### CODE REVIEW

**Comment by peppy** [comment]
haven't looked at the code, but please try and test the most *common* ways to break a new feature before submitting for review

screen shot 2017-06-11 at 16 54 28

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 1]
```diff
@@ -190,6 +194,46 @@ protected MenuOverlay()
...
         }
 
+        protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
+        {
+            if (args.Repeat) return false;
+
+            var buttonsAmount = Buttons.Children.Count();
+            if (buttonsAmount == 0) return false;
+
+            foreach (var button in Buttons.Children)
```
Comment (MrTheMake) [inline osu.Game/Screens/Play/MenuOverlay.cs]: This should only be done when pressing UP or DOWN.

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 2]
```diff
@@ -190,6 +194,46 @@ protected MenuOverlay()
...
+                button.TriggerOnHoverLost(state);
+
+            switch (args.Key)
+            {
+                case Key.Down:
+                    if (selectedButton == 0 || selectedButton == buttonsAmount)
+                        selectedButton = 1;
```
Comment (MrTheMake) [inline osu.Game/Screens/Play/MenuOverlay.cs]: I personally don't like the representation of "no button is selected" as 0. I think it is more intuitive to use -1.

Comment (Jorolf) [inline osu.Game/Screens/Play/MenuOverlay.cs]: 0 is generally the start of an array/list so it is confusing to make 1 the first button

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 3]
```diff
@@ -190,6 +194,46 @@ protected MenuOverlay()
...
+                    else
+                        selectedButton++;
+                    Buttons.Children.ElementAt(selectedButton - 1).TriggerOnHover(state);
```
Comment (Jorolf) [inline osu.Game/Screens/Play/MenuOverlay.cs]: this would remove the -1 here as well

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 4]
```diff
@@ -184,20 +190,136 @@ private void load(OsuColour colours)
...
+            switch (args.Key)
+            {
+                case Key.Down:
+                    if (buttonsAmount == 0) return false;
+
+                    if (selectedButton != -1)
+                        Buttons.Children.ElementAt(selectedButton - 1).Deselect(state);
+
+                    if (selectedButton == -1 || selectedButton == buttonsAmount)
+                        selectedButton = 1;
```
Comment (MrTheMake) [inline osu.Game/Screens/Play/MenuOverlay.cs]: This should be changed to 0 now.

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 5]
```diff
@@ -184,20 +190,136 @@ private void load(OsuColour colours)
...
+
+            switch (args.Key)
+            {
+                case Key.Down:
+                    if (buttonsAmount == 0) return false;
+
+                    if (selectedButton != -1)
+                        Buttons.Children.ElementAt(selectedButton - 1).Deselect(state);
+
+                    if (selectedButton == -1 || selectedButton == buttonsAmount)
```
Comment (MrTheMake) [inline osu.Game/Screens/Play/MenuOverlay.cs]: buttonsAmount - 1

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 6]
```diff
@@ -184,20 +190,136 @@ private void load(OsuColour colours)
...
+                        selectedButton = 1;
+                    else
+                        selectedButton++;
+                    Buttons.Children.ElementAt(selectedButton - 1).Select(state);
```
Comment (MrTheMake) [inline osu.Game/Screens/Play/MenuOverlay.cs]: ...ElementAt(selectedButton)...

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 7]
```diff
@@ -184,20 +190,136 @@ private void load(OsuColour colours)
...
+                    return true;
+
+                case Key.Up:
+                    if (buttonsAmount == 0) return false;
+
+                    if (selectedButton != -1)
+                        Buttons.Children.ElementAt(selectedButton - 1).Deselect(state);
```
Comment (MrTheMake) [inline osu.Game/Screens/Play/MenuOverlay.cs]: ...ElementAt(selectedButton)...

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 8]
```diff
@@ -184,20 +190,136 @@ private void load(OsuColour colours)
...
+
+                    if (selectedButton == -1 || selectedButton == 1)
```
Comment (MrTheMake) [inline osu.Game/Screens/Play/MenuOverlay.cs]: || selectedButton == 0

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 9]
```diff
@@ -184,20 +190,136 @@ private void load(OsuColour colours)
...
+                        selectedButton = buttonsAmount;
+                    else
+                        selectedButton--;
+                    Buttons.Children.ElementAt(selectedButton - 1).Select(state);
```
Comment (MrTheMake) [inline osu.Game/Screens/Play/MenuOverlay.cs]: ...ElementAt(selectedButton)...

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 10]
```diff
@@ -184,20 +190,136 @@ private void load(OsuColour colours)
...
+                    return true;
+
+                case Key.Enter:
+                    if (selectedButton != -1)
+                    {
+                        Buttons.Children.ElementAt(selectedButton - 1).TriggerOnClick();
```
Comment (MrTheMake) [inline osu.Game/Screens/Play/MenuOverlay.cs]: ...ElementAt(selectedButton)...

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 11]
```diff
@@ -184,20 +190,136 @@ private void load(OsuColour colours)
...
+        {
+            if (args.Repeat) return false;
+
+            switch (args.Key)
+            {
+                case Key.Down:
+                    if (buttonsAmount == 0) return false;
+
+                    if (selectedButton != -1)
+                        Buttons.Children.ElementAt(selectedButton - 1).Deselect(state);
```
Comment (MrTheMake) [inline osu.Game/Screens/Play/MenuOverlay.cs]: ...ElementAt(selectedButton)...

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 12]
```diff
@@ -184,20 +190,137 @@ private void load(OsuColour colours)
...
+            state = manager.CurrentState;
+            header.Colour = colours.Yellow;
+        }
+
+        private int buttonsAmount => Buttons.Children.Count();
+        private int selectedButton = -1;
+
+        protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
+        {
+            if (args.Repeat) return false;
```
Comment (Jorolf) [inline osu.Game/Screens/Play/MenuOverlay.cs]: return true, otherwise the volume overlay shows

File: osu.Game/Screens/Play/MenuOverlay.cs [hunk 13]
```diff
@@ -184,20 +190,139 @@ private void load(OsuColour colours)
...
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours, UserInputManager manager)
         {
-            AlwaysReceiveInput = true;
-            RelativeSizeAxes = Axes.Both;
+            state = manager.CurrentState;
+            header.Colour = colours.Yellow;
+        }
+
+        private int buttonsAmount => Buttons.Children.Count();
```
Comment (smoogipooo) [inline osu.Game/Screens/Play/MenuOverlay.cs]: `buttonCount`

**Review by smoogipoo — CHANGES_REQUESTED** [review]
I'd say this is looking okay on an initial pass.

**Comment by EVAST9919** [comment]
Don't review this yet, some mistakes has been founded.

**Comment by EVAST9919** [comment]
Main issue for now: button which has been selected and clicked by keyboard can't be deselected by selecting another button for the first time.

**Comment by EVAST9919** [comment]
Closing untill I found a better way to implement this

**Pull request closed without merging**